File: bilateral_filter.py
import numpy as np
from PIL import Image
import sys
import scipy.signal


def gaussian(x, sigma):
    return (1.0 / (2 * np.pi * (sigma ** 2))) * np.exp(- (x ** 2) / (2 * sigma ** 2))


def bilateral_filter(source, radius, std_i, std_s):
    filtered_image = np.zeros_like(source).astype(float)
    W = 0

    pad = np.pad(source, (radius, radius), mode="symmetric")

    for i in range(-radius, radius):
        for j in range(-radius, radius):

            neighbour = pad[radius + i: radius + i + source.shape[0],
                            radius + j: radius + j + source.shape[1]]

            distance_x = (i)**2
            distance_y = (j)**2
            distance = np.sqrt(distance_x + distance_y)

            gi = gaussian((neighbour - source), std_i)
            gs = gaussian(distance, std_s)

            w = gi * gs
            W += w
            filtered_image += neighbour * w

    return np.around(filtered_image / W).astype(np.uint8)


if __name__ == "__main__":

    src = cv2.imread(str(sys.argv[1]), 0)
    dest = "filtered_image_own.png"

    with Image.open(src) as pic:

        pic = np.array(pic)

        filtered_image = bilateral_filter(pic, 5, 1, 3)

        with Image.fromarray(filtered_image) as output:
            output.save(dest, mode="L")

File: bilateral_filter.cpp
#include <iostream>
#include <stdio.h>
#include <opencv2/opencv.hpp>

using namespace cv;

float distance(int x, int y, int i, int j) {
    return float(sqrt(pow(x - i, 2) + pow(y - j, 2)));
}

double gaussian(float x, double sigma) {
    return exp(-(pow(x, 2))/(2 * pow(sigma, 2))) / (2 * CV_PI * pow(sigma, 2));

}

void applyBilateralFilter(Mat source, Mat filteredImage, int x, int y, int diameter, double sigmaI, double sigmaS) {
    double iFiltered = 0;
    double wP = 0;
    int neighbor_x = 0;
    int neighbor_y = 0;
    int half = diameter / 2;

    for(int i = 0; i < diameter; i++) {
        for(int j = 0; j < diameter; j++) {
            neighbor_x = x - (half - i);
            neighbor_y = y - (half - j);
            double gi = gaussian(source.at<uchar>(neighbor_x, neighbor_y) - source.at<uchar>(x, y), sigmaI);
            double gs = gaussian(distance(x, y, neighbor_x, neighbor_y), sigmaS);
            double w = gi * gs;
            iFiltered = iFiltered + source.at<uchar>(neighbor_x, neighbor_y) * w;
            wP = wP + w;
        }
    }
    iFiltered = iFiltered / wP;
    filteredImage.at<double>(x, y) = iFiltered;


}

Mat bilateralFilterOwn(Mat source, int diameter, double sigmaI, double sigmaS) {
    Mat filteredImage = Mat::zeros(source.rows,source.cols,CV_64F);
    int width = source.cols;
    int height = source.rows;

    for(int i = 2; i < height - 2; i++) {
        for(int j = 2; j < width - 2; j++) {
            applyBilateralFilter(source, filteredImage, i, j, diameter, sigmaI, sigmaS);
        }
    }
    return filteredImage;
}


int main(int argc, char** argv ) {
    Mat src;
    src = imread( argv[1], 0 );
    imwrite("original_image_grayscale.png", src);

    if ( !src.data )
    {
        printf("No image data \n");
        return -1;
    }

    Mat filteredImageOpenCV;
    bilateralFilter(src, filteredImageOpenCV, 5, 12.0, 16.0);
    imwrite("filtered_image_OpenCV.png", filteredImageOpenCV);
    Mat filteredImageOwn = bilateralFilterOwn(src, 5, 12.0, 16.0);
    imwrite("filtered_image_own.png", filteredImageOwn);

    return 0;
}